src/test/scala/com/microsoft/hyperspace/index/E2EHyperspaceRulesTest.scala

```diff
@@ -670,13 +670,245 @@ class E2EHyperspaceRulesTest extends QueryTest with HyperspaceSuite {
+        def query(): DataFrame = df.filter("c3 == 'facebook'").select("c3", "c4")
+
+        val indexManager = Hyperspace
+          .getContext(spark)
+          .indexCollectionManager
+        val indexEntry =
+          indexManager.getIndexes().filter(_.name.equals(indexConfig.indexName)).head
+        assert(indexEntry.getTagValue(IndexLogEntryTags.INMEMORYFILEINDEX_INDEX_ONLY).isEmpty)
+        query().collect
+        val location1 = query().queryExecution.optimizedPlan.collect {
+          case LogicalRelation(HadoopFsRelation(loc, _, _, _, _, _), _, _, _) =>
+            loc
+        }.head
+
+        assert(indexEntry.getTagValue(IndexLogEntryTags.INMEMORYFILEINDEX_INDEX_ONLY).nonEmpty)
+        val location2 = query().queryExecution.optimizedPlan.collect {
+          case LogicalRelation(HadoopFsRelation(loc, _, _, _, _, _), _, _, _) =>
+            loc
+        }.head
+
+        assert(location2.eq(location1))
+        assert(
+          location2.eq(
+            indexEntry.getTagValue(IndexLogEntryTags.INMEMORYFILEINDEX_INDEX_ONLY).get))
+      }
+    }
+  }
+
+  test("Verify InMemoryFileIndex cache is not used for a different plan with Hybrid scan.") {
+    withTempPathAsString { testPath =>
+      val indexConfig = IndexConfig("index", Seq("c3"), Seq("c4"))
+      import spark.implicits._
+      SampleData.testData
+        .toDF("c1", "c2", "c3", "c4", "c5")
+        .limit(10)
+        .write
+        .parquet(testPath)
+      val df = spark.read.load(testPath)
+      hyperspace.createIndex(df, indexConfig)
+      withIndex(indexConfig.indexName) {
+        spark.enableHyperspace()
+        val df = spark.read.parquet(testPath)
+        def query(df: DataFrame): DataFrame = {
+          df.filter("c3 == 'facebook'").select("c3", "c4")
+        }
+        def getQueryPlanKey(df: DataFrame): LogicalPlan = {
+          spark.disableHyperspace()
+          val p = query(df).queryExecution.optimizedPlan
+          spark.enableHyperspace()
+          p
+        }
+        def getFsLocation(plan: LogicalPlan): FileIndex = {
+          plan.collect {
+            case LogicalRelation(HadoopFsRelation(loc, _, _, _, _, _), _, _, _) =>
+              loc
+          }.head
+        }
+
+        val indexManager = Hyperspace
+          .getContext(spark)
+          .indexCollectionManager
+        val indexEntry =
+          indexManager.getIndexes().filter(_.name.equals(indexConfig.indexName)).head
+        assert(indexEntry.getTagValue(IndexLogEntryTags.INMEMORYFILEINDEX_INDEX_ONLY).isEmpty)
+        query(df).collect
+        val location1 = getFsLocation(query(df).queryExecution.optimizedPlan)
+        assert(indexEntry.getTagValue(IndexLogEntryTags.INMEMORYFILEINDEX_INDEX_ONLY).nonEmpty)
+
+        SampleData.testData
+          .toDF("c1", "c2", "c3", "c4", "c5")
+          .limit(3)
+          .write
+          .mode("append")
+          .parquet(testPath)
+
+        val df2 = spark.read.parquet(testPath)
```

Comment:
Can you add short comments describing each test scenario? A bit hard to follow...